File: Plane3D.java
import static java.lang.Math.*;

/** represents a 3D plane
 * @author Jordan Lau 340600240 */
public class Plane3D {

	/** x coefficient of the plane equation in cartesian form */
	protected double a;
	/** y coefficient of the plane equation in cartesian form */
	protected double b;
	/** z coefficient of the plane equation in cartesian form */
	protected double c;
	/** scalar of the plane equation in cartesian form */
	protected double d;

	/** Constructor from 3 points
	 * @param p1
	 * @param p2
	 * @param p3 */
	public Plane3D(Point3D p1, Point3D p2, Point3D p3){
		//normal vector = cross product of direction vectors v1 and v2
		Point3D v1 = new Point3D(p1.getX()-p2.getX(), p1.getY()-p2.getY(), p1.getZ()-p2.getZ());
		Point3D v2 = new Point3D(p2.getX()-p3.getX(), p2.getY()-p3.getY(), p2.getZ()-p3.getZ());
		a = v1.getY()*v2.getZ() - v2.getY()*v1.getZ();
		b = v1.getZ()*v2.getX() - v2.getZ()*v1.getX();
		c = v1.getX()*v2.getY() - v2.getX()*v1.getY();
		d = -1*a*p1.getX() -1*b*p1.getY() -1*c*p1.getZ();
	}
	
	/** Constructor from 4 parameters
	 * @param a x coefficient
	 * @param b y coefficient
	 * @param c z coefficient
	 * @param d scalar */
	public Plane3D(double a, double b, double c, double d){
		this.a = a;
		this.b = b;
		this.c = c;
		this.d = d;
	}
	
	/** computes the euclidean distance between a point and the plane
	 * @param pt the point from which the distance to this Point object */
	public double getDistance(Point3D pt){
		double numerator, denominator;
		numerator = abs(a*pt.getX() + b*pt.getY() + c*pt.getZ() + d);
		denominator = sqrt(pow(a,2) + pow(b,2) + pow(c,2));
		return numerator/denominator;
	}

}

File: Plane3D.go
package main

import . "math"

type Plane3D struct {
    A float64
    B float64
    C float64
    D float64
}

// computes the plane defined by a slice of 3 points
func GetPlane(points [3]Point3D) Plane3D{
    p1 := points[0]
    p2 := points[1]
    p3 := points[2]

    //direction vectors v1 and v2
    v1 := Point3D{p1.X-p2.X, p1.Y-p2.Y, p1.Z-p2.Z}
    v2 := Point3D{p2.X-p3.X, p2.Y-p3.Y, p2.Z-p3.Z}

    //cross product of v1 and v2 (to get normal vector)
    a := v1.Y*v2.Z - v2.Y*v1.Z
    b := v1.Z*v2.X - v2.Z*v1.X
    c := v1.X*v2.Y - v2.X*v1.Y
    
    //solve for last variable
    d := -1*a*p1.X -1*b*p1.Y -1*c*p1.Z
    
    return Plane3D{a,b,c,d}
}

// I copied this helper function from
// https://stackoverflow.com/questions/37334119/how-to-delete-an-element-from-a-slice-in-golang
func remove(slice []Point3D, s int) []Point3D {
    return append(slice[:s], slice[s+1:]...)
}

// creates a new slice of points in which all points belonging to the plane have been removed
func RemovePlane(plane Plane3D, points []Point3D, eps float64) []Point3D{
    newPoints := make([]Point3D,0)
    for i, point := range(newPoints){
        if (plane.GetDistance(point) < eps){
            newPoints = remove(newPoints, i)
        }
    }
    return newPoints
}

/** computes the euclidean distance between a point and the plane*/
func (plane Plane3D) GetDistance(pt Point3D) float64{
    numerator := Abs(plane.A*pt.X + plane.B*pt.Y + plane.C*pt.Z + plane.D)
    denominator := Sqrt(Pow(plane.A,2) + Pow(plane.B,2) + Pow(plane.C,2))
    return numerator/denominator
}